### PR TITLE
refactor(input-service): make it more pluggable and separated

### PR DESCRIPTION
**Pull request by connor4312**
Just this morning's refactoring.

I split out the InputService, which previously had keyboard and gamepad stuff amalgamated into a single class, into pluggable and more testable input method classes, one for the gamepad, and one for the keyboard. I also reworked them to be pure classes that simply expose an rxjs observable of buttons they output, without any state stored on the class itself.

This means the focus service is super simple, it just merges in all registered strategies into a single rxjs stream. It also means that getting a stream of buttons you want to deal with in the consuming application is a whole lot easier, you just subscribe to the stream.

Next up this afternoon will be writing some tests and then giving a similar treatment to the focus service 🙂

### CODE REVIEW

File: src/components/util.test.tsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import * as React from 'react';
+
+const container = document.createElement('div');
+document.body.appendChild(container);
+
+let mountings: ReactWrapper[] = [];
+
+/**
+ * mountToDOM is Enzyme's mount function that attaches the element to the
+ * document, which Enzyme doesn't do by default. This is needed for many
```
Comment (irbrownie) [inline src/components/util.test.tsx:14]: enzyme or mocha?

File: src/focus/is-for-form.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  // Deal with the output ourselves, allowing arcade-machine to handle it only
+  // if the key press would not have any effect in the context of the input.
+  const input = selected as HTMLInputElement | HTMLTextAreaElement;
+  const { type } = input;
+  if (
+    type !== 'text' &&
+    type !== 'search' &&
+    type !== 'url' &&
+    type !== 'tel' &&
+    type !== 'password'
```
Comment (irbrownie) [inline src/focus/is-for-form.ts]: how are number inputs handled?

File: src/components/util.test.tsx (hunk 1 above)
Comment (connor4312) [reply src/components/util.test.tsx:14]: Enzyme

File: src/input/gamepad-input.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    navigator.gamepadInputEmulation = 'keyboard';
+    return true;
+  }
+
+  /**
+   * Checks for input provided by the gamepad and fires off events as
+   * necessary. It schedules itself again provided that there's still
+   * a connected gamepad somewhere.
+   */
+  private pollGamepad(now: number, gamepads: { [id: string]: IGamepadWrapper }): Button | null {
```
Comment (irbrownie) [inline src/input/gamepad-input.ts:107]: What's the polling based off here?

Comment (connor4312) [reply src/input/gamepad-input.ts:107]: requestAnimationFrame

File: src/input/keyboard-input.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  ]);
+
+  public readonly observe = fromEvent<KeyboardEvent>(window, 'keydown').pipe<Button>(
+    map(ev => {
+      const direction = KeyboardInput.codeDirectionMap.get(ev.keyCode);
+      if (direction === undefined) {
+        return undefined;
+      }
+
+      ev.preventDefault();
```
Comment (irbrownie) [inline src/input/keyboard-input.ts]: small nit: I'd throw preventDefault in a tap.

File: src/focus-service.ts [hunk 5]
```diff
@@ -111,75 +58,21 @@ export class FocusService {
...
-  }
-
-  /**
-   * Useful for resetting all focus traps e.g. on page navigation
-   */
-  public clearAllTraps() {
-    this.focusStack.length = 0;
-    this.focusRoot = defaultFocusRoot;
+  private get selected() {
+    return document.activeElement as HTMLElement;
```
Comment (saadataz) [inline src/focus-service.ts]: we wanna keep `document.activeElement` separate from `focus.selected` to support showing the elements as focused without actually focusing them. This is necessary for smooth scrolling to work

Comment (connor4312) [reply src/focus-service.ts]: @saadataz it looks like there's a (new?) [preventScroll option](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/focus) of the .focus() method. That seems to fill the use case of smooth scroll.

**Pull request merged**